iml-request-retry/examples/demo-server-client.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use iml_request_retry::{retry_future, retry_future_gen, RetryAction};
+use reqwest::{Client, Response, Url};
+use serde_derive::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::io;
+use std::net::SocketAddr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::time::Duration;
+use tokio::runtime::Runtime;
+use tokio::timer::delay_for;
+use warp::http::StatusCode;
+use warp::Filter;
+
+static GLOBAL_SERVER_COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+/// A serialized message to report in JSON format.
+#[derive(Serialize, Deserialize, Debug)]
+struct Hello {
+    client: u32,
+    counter: Option<usize>,
+}
+
+fn main() {
```

Comment:
With tokio 0.2 you can just make the main fn async which will simplify the nesting etc:

```suggestion
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
```